engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,575 @@ package ingestion
-		Uint64("stopHeight", height).
-		Bool("crash", crash).
-		Uint64("old_height", oldHeight).
-		Bool("old_crash", oldCrash).
-		Msg("new stopHeight set")
+	if s.stopBoundary == nil {
+		// if there is no stop boundary set, we can set it to anything
+		return true, ""
+	}
 
-	s.state = StopControlSet
+	if s.stopBoundary.cannotBeChanged {
+		return false, fmt.Sprintf(
+			"cannot update stopHeight, stopping commenced for %s",
+			s.stopBoundary,
+		)
+	}
 
-	s.stopHeight = height
-	s.crash = crash
-	s.stopAfterExecuting = flow.ZeroID
+	if s.stopBoundary.fromVersionBeacon != fromVersionBeacon &&
+		newHeight > s.stopBoundary.StopHeight {
+		// if one stop was set by the version beacon and the other one was manual
+		// we can only update if the new stop is earlier
+
+		// this prevents users moving the stopHeight forward when a version boundary
+		// is earlier, and prevents version beacons from moving the stopHeight forward
+		// when a manual stop is earlier.
+		return false, "cannot update stopHeight, new stop height is later than the " +
+			"current one (or removing a stop)"
+	}
 
-	return oldHeight, oldCrash, nil
+	return true, ""
 }
 
-// GetStopHeight returns:
-//   - stopHeight
-//   - crash
-//
-// Values are undefined if they were not previously set
-func (s *StopControl) GetStopHeight() (uint64, bool) {
+// GetStop returns the upcoming stop parameters or nil if no stop is set.
+func (s *StopControl) GetStop() *StopParameters {
 	s.RLock()
 	defer s.RUnlock()
 
-	return s.stopHeight, s.crash
+	if s.stopBoundary == nil {
+		return nil
+	}
+
+	p := s.stopBoundary.StopParameters
+	return &p
 }
 
-// blockProcessable should be called when new block is processable.
+// BlockProcessable should be called when new block is processable.
 // It returns boolean indicating if the block should be processed.
-func (s *StopControl) blockProcessable(b *flow.Header) bool {
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
 	s.Lock()
 	defer s.Unlock()
 
-	if s.state == StopControlOff {
-		return true
+	// don't process anymore blocks if stopped
+	if s.stopped {
+		return false
 	}
 
-	if s.state == StopControlPaused {
-		return false
+	// if no stop is set process all blocks
+	if s.stopBoundary == nil {
+		return true
 	}
 
-	// skips blocks at or above requested stopHeight
-	if b.Height >= s.stopHeight {
+	// Skips blocks at or above requested stopHeight
+	// doing so means we have started the stopping process
+	if b.Height >= s.stopBoundary.StopHeight {
 		s.log.Warn().
-			Int8("previous_state", int8(s.state)).
-			Int8("new_state", int8(StopControlCommenced)).
 			Msgf(
 				"Skipping execution of %s at height %d"+
-					" because stop has been requested at height %d",
+					" because stop has been requested %s",
 				b.ID(),
 				b.Height,
-				s.stopHeight,
+				s.stopBoundary,
 			)
 
-		s.state = StopControlCommenced // if block was skipped, move into commenced state
+		s.stopBoundary.cannotBeChanged = true
```

Comment:
what about
```suggestion
		s.stopBoundary.changable = false
```

Reply:
I wanted something that has default of false. That is why the naming is weird.